bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/TypeChecker.java

```diff
@@ -467,8 +467,30 @@ public static boolean isReferenceEqual(Object lhsValue, Object rhsValue) {
         Type lhsType = getType(lhsValue);
         Type rhsType = getType(rhsValue);
 
-        if (isSimpleBasicType(lhsType) && isSimpleBasicType(rhsType)) {
-            return isEqual(lhsValue, rhsValue);
+        switch(lhsType.getTag()) {
+            case TypeTags.INT_TAG:
+                if (rhsType.getTag() != TypeTags.BYTE_TAG || rhsType.getTag() != TypeTags.INT_TAG) {
```

Comment:
What happen if we have integer sub types

Reply:
what is the difference between isReferenceEqual and isEqual for simple basic types?

Reply:
We have used `getType` method to get the types of values. So the types of integer sub types also will be given as the type integer. Equality and exact equality checks are not same for `float` and `decimal` types. So we can't simply use `isEqual()` check of simple basic types for `isReferenceEqual()` check. Or else we need to exclude the `float` and `decimal` cases and we can handle them separately.

Reply:
Lets handle float and decimal diffrently and other types use same isEqual which will minimise code duplication.